icicle/backend/cpu/src/hash/cpu_merkle_tree.cpp

```diff
@@ -40,42 +40,41 @@ namespace icicle {
 
         if (0 < layer_idx) {
           // update nof_hashes to the next layer (below)
-          const uint64_t next_layer_total_size = layer_hashes[layer_idx - 1].output_size();
-          nof_hashes = nof_hashes * cur_layer.m_hash.default_input_chunk_size() / next_layer_total_size;
+          const uint64_t cur_layer_input_size = layer_hashes[layer_idx].default_input_chunk_size();
+          nof_hashes = nof_hashes * cur_layer_input_size / layer_hashes[layer_idx - 1].output_size();
 
           // Calculate path_size
-          m_pruned_path_size +=
-            layer_hashes[layer_idx].default_input_chunk_size() - layer_hashes[layer_idx - 1].output_size();
-          m_full_path_size += layer_hashes[layer_idx].default_input_chunk_size();
+          m_pruned_path_size += cur_layer_input_size - layer_hashes[layer_idx - 1].output_size();
+          m_full_path_size += cur_layer_input_size;
         }
       }
     }
 
-    // TODO: handle size
+    ~CPUMerkleTreeBackend()
+    {
+      for (auto& pair : m_map_segment_id_2_inputs) {
+        delete pair.second; // Ensure all dynamically allocated memory is freed
```

Comment:
modern C++ should avoid new/malloc+delete/free and use smart pointers anyway. This would avoid the leak and also works when exceptions are thrown.
There is usually no reason to use malloc/new

Reply:
fixed